File: sbe/binance/spot_3_2/generated-python/spot_sbe/ExternalLockUpdateEvent.py
"""Generated SBE (Simple Binary Encoding) message codec."""

import struct
from typing import List, Optional, Tuple
from io import BytesIO

class ExternalLockUpdateEvent:
    """SBE message: ExternalLockUpdateEvent."""

    TEMPLATE_ID = 604
    SCHEMA_ID = 3
    SCHEMA_VERSION = 2
    BLOCK_LENGTH = 27

    def __init__(self):
        self.event_time = None
        self.clear_time = None
        self.qty_exponent = None
        self.locked_qty_delta = None
        self.subscription_id = None
        self.asset = b''

    def encode(self) -> bytes:
        """Encode the message to bytes."""
        buffer = BytesIO()

        if self.event_time is not None:
            buffer.write(struct.pack('<q', self.event_time))
        if self.clear_time is not None:
            buffer.write(struct.pack('<q', self.clear_time))
        if self.qty_exponent is not None:
            buffer.write(struct.pack('<b', self.qty_exponent))
        if self.locked_qty_delta is not None:
            buffer.write(struct.pack('<q', self.locked_qty_delta))
        if self.subscription_id is not None:
            buffer.write(struct.pack('<H', self.subscription_id))

        return buffer.getvalue()

    def decode(self, data: bytes) -> None:
        """Decode the message from bytes."""
        buffer = BytesIO(data)

        self.event_time = struct.unpack('<q', buffer.read(8))[0]
        self.clear_time = struct.unpack('<q', buffer.read(8))[0]
        self.qty_exponent = struct.unpack('<b', buffer.read(1))[0]
        self.locked_qty_delta = struct.unpack('<q', buffer.read(8))[0]
        self.subscription_id = struct.unpack('<H', buffer.read(2))[0]

File: sbe/binance/spot_3_1/generated-php/spot_sbe/ExternalLockUpdateEvent.php
<?php
/**
 * Generated SBE (Simple Binary Encoding) message codec.
 */

class ExternalLockUpdateEvent
{
    public const TEMPLATE_ID = 604;
    public const SCHEMA_ID = 3;
    public const SCHEMA_VERSION = 1;
    public const BLOCK_LENGTH = 27;

    public int|float|array|null $eventTime = null;
    public int|float|array|null $clearTime = null;
    public int|float|array|null $qtyExponent = null;
    public int|float|array|null $lockedQtyDelta = null;
    public int|float|array|null $subscriptionId = null;
    public string $asset = '';

    private function decodeVarData(string $data, int &$offset): string
    {
        $length = unpack('V', substr($data, $offset, 4))[1];
        $offset += 4;
        $value = substr($data, $offset, $length);
        $offset += $length;
        return $value;
    }

    public function encode(): string
    {
        $buffer = '';

        if ($this->eventTime !== null) {
            $buffer .= pack('q', $this->eventTime);
        }
        if ($this->clearTime !== null) {
            $buffer .= pack('q', $this->clearTime);
        }
        if ($this->qtyExponent !== null) {
            $buffer .= pack('c', $this->qtyExponent);
        }
        if ($this->lockedQtyDelta !== null) {
            $buffer .= pack('q', $this->lockedQtyDelta);
        }
        if ($this->subscriptionId !== null) {
            $buffer .= pack('v', $this->subscriptionId);
        }

        return $buffer;
    }

    public function decode(string $data): void
    {
        $offset = 0;

        $this->eventTime = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->clearTime = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->qtyExponent = unpack('c', substr($data, $offset, 1))[1];
        $offset += 1;
        $this->lockedQtyDelta = unpack('q', substr($data, $offset, 8))[1];
        $offset += 8;
        $this->subscriptionId = unpack('v', substr($data, $offset, 2))[1];
        $offset += 2;

        // Skip to end of block for forward compatibility
        $offset = 27;


        $this->asset = $this->decodeVarData($data, $offset);
    }
}
